src/httpd_artwork.c

```diff
@@ -0,0 +1,293 @@
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#ifdef HAVE_CONFIG_H
+# include <config.h>
+#endif
+
+#include <regex.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "httpd_artwork.h"
+#include "db.h"
+#include "library.h"
+#include "logger.h"
+#include "misc.h"
+#include "player.h"
+#include "artwork.h"
+
+
+static int
+artwork_to_evbuf(struct httpd_request *hreq, struct db_group_info *dbgri, uint32_t *dbmfi_id)
+{
+  int ret = 0;
+  int len = 0;
+  const char* ctype = "invalid";
+  char clen[32];
+  struct evkeyvalq *headers;
+
+  if ( (dbgri == NULL && dbmfi_id == NULL) || (dbgri && dbmfi_id)) 
+    {
+      return -1;
+    }
+
+  if (dbgri)
+    {
+      int  gpid;
+      safe_atoi32(dbgri->id, &gpid);
+      if ( (ret = artwork_get_group(hreq->reply, gpid, 600, 600)) < 0)
+        {
+          DPRINTF(E_LOG, L_WEB, "artwork: failed to retreive for group %d\n", gpid);
+        }
+    }
+  if (dbmfi_id)
+    {
+      if ( (ret = artwork_get_item(hreq->reply, *dbmfi_id, 600, 600)) < 0)
+        {
+          DPRINTF(E_LOG, L_WEB, "artwork: failed to retreive for item %d\n", *dbmfi_id);
+        }
+    }
+
+  len = evbuffer_get_length(hreq->reply);
+  switch (ret)
+    {
+      case ART_FMT_PNG:
+        ctype = "image/png";
+        break;
+
+      case ART_FMT_JPEG:
+        ctype = "image/jpeg";
+        break;
+
+      default:
+        if (len > 0) evbuffer_drain(hreq->reply, len);
+        ret = -1;
+    }
+  DPRINTF(E_DBG, L_WEB, "artwork: via %s  type=%s len=%d\n", (dbgri ? "group" : "item"), ctype, len);
+
+  if (ret)
+    {
+      headers = evhttp_request_get_output_headers(hreq->req);
+      evhttp_remove_header(headers, "Content-Type");
+      evhttp_add_header(headers, "Content-Type", ctype);
+      snprintf(clen, sizeof(clen), "%ld", (long)len);
+      evhttp_add_header(headers, "Content-Length", clen);
+
+      ret = 0;
+    }
+  return ret;
+}
+
+static int
+fetch_artwork(struct httpd_request *hreq, const char *artist_id, const char *album_id)
+{
+  struct db_group_info dbgri;
+  struct query_params query_params;
+  int ret;
+
+  if (artist_id == NULL && album_id == NULL || (artist_id && album_id)) 
+    {
+      return -1;
+    }
+  memset(&query_params, 0, sizeof(struct query_params));
+  memset(&dbgri, 0, sizeof(dbgri));
```

Comment:
Use sizeof(struct ...), it protects against errors if you or someone else one day changes dbgri to a pointer.